### ISSUE TITLE
Ivy: default input should not stick after typing

### ISSUE THREAD

**Issue by PythonNut** [abo-abo/swiper#231]
`ivy` detects the default match based on the text around point. However, as the user begins typing the default match remains selected, rather than the first one.

This is important for non-default sorting options and #207, where the sorting can change the candidate, leaving the user with a random non-initial candidate selected instead.

**Comment by abo-abo** [comment]
This should be take care of once any kind of sorting is added. Currently, there is none. In some situations, the default match is prepended to the collection.  And the matches are selected in the same order as they are in the collection.

**Comment by PythonNut** [comment]
> This should be take care of once any kind of sorting is added. 

Do you mean by added by users to their configs or by you as a feature?

**Comment by abo-abo** [comment]
As a feature, for example flx sorting.

**Comment by PythonNut** [comment]
@abo-abo any idea how I can fix this for #207? Right now, it's the most important usage-problem for the current `flx`-sorting work.

**Comment by tsdh** [comment]
I was just going to report a problem which is probably a different incarnation of this issue.  Namely, I use `counsel-M-x` and have installed the `smex` package so that the commands are sorted by frequency.  Now assume I want to invoke `list-packages`.  At the `M-x` prompt, I get this initial list:

```
magit-status <<<
list-packages
gnus
...
```

where the `<<<` shows what's selected.  If I now type `lif`, i.e., I make a typo, then the list becomes just:

```
life <<<
```

Now I see my typo and delete the `f` from the input and get these completions:

```
list-packages
list-colors-display
life <<<
lisp-mode
...
```

As you can see, the selection stuck to the `life` entry whereas I would expect to have the first matching entry selected again.  I.e., as long as the user changes the input, there is no value in sticking to the previously selected entry if possible.  If that was the right one, the user wouldn't type anymore.

Is that the same issue or should I open a separate one for this?

**Comment by abo-abo** [comment]
> I.e., as long as the user changes the input, there is no value in sticking to the previously selected entry if that's still a valid candidate.

The code is currently written in a straightforward way, which doesn't remember the state of completion at the previous step. In this case, once a character is erased, the available variant to select is either the first candidate or the current one (which is better since it avoids jumping).

I think that writing it in a way that remembers the complete history of completion could be hard.  For example, let's say I entered 2 chars, pasted 10, and erased 5 at once. What should the state be?

If you have an idea of doing this in a sane way, please contribute. I'll try to at least merge it as a customizable option - I'm OK with the way it works now.

**Comment by tsdh** [comment]
I didn't suggest to keep track of states rather I suggested to select the first and not the candidate which has been selected previously. Yes, then selection jumps in case you delete a chat but IMO that's OK given that the user has control over sorting and thus can sort the important matches on top of the list (like smex does by command frequency).

**Comment by abo-abo** [comment]
> I suggested to select the first and not the candidate which has been selected previously

This will make `swiper` work worse: when I delete a char I don't want a scroll in the main window.  I do this quite often: deleting a char/word doesn't mean that I made an error, it could be that I want to alter the query to match the current line, but also other lines.  This also applies to `counsel-git-grep` and `counsel-ag`.

The change that you want is as easy as commenting out `(when (and tail ivy--old-cands`. But I want to keep this behavior for at least the three commands that I mentioned above, maybe more.  Anyway, have a look there, maybe add a `defcustom` that makes it work the way you want.

**Comment by tsdh** [comment]
Ok, you are probably right with respect to swiper. I'll see if there's anything that can be done when I find some spare time.

**Comment by PythonNut** [comment]
@abo-abo 

> This will make `swiper` work worse: when I delete a char I don't want a scroll in the main window. I do this quite often: deleting a char/word doesn't mean that I made an error, it could be that I want to alter the query to match the current line, but also other lines. This also applies to `counsel-git-grep` and `counsel-ag`.

Doesn't this sound like a feature that should be enabled for the specific commands where it applies. It's more of a special case than a reasonable general default.

**Comment by PythonNut** [comment]
@abo-abo 

Okay, so I merged my `feature/ivy` branch into my personal config after #207 was resolved, and the sorting doesn't play very well with this issue. As far as I can tell, once this is resolved, `ivy` will have gained feature parity with my old `ido`+`ido-ubiquitous`+`ido-vertical-mode`+`smex`+`flx-ido` setup.

**Comment by abo-abo** [comment]
> As far as I can tell, once this is resolved

Could you provide a concrete example for me to work on? Something test-able like this snippet from ivy-test.el:

```elisp
(should (equal
         (ivy-with '(ivy-read "pattern: " '("blue" "yellow"))
                   "y C-m")
         "yellow"))
```

**Comment by PythonNut** [comment]
@abo-abo great idea.

I haven't actually run these tests. I hope they're valid.

```emacs
(should (equal
         (ivy-with '(ivy-read "pattern: " '("blue" "yellow"))
                   "y <backspace> C-m")
         "blue"))

;; requires flx sorting
(should (equal
         (ivy-with '(let ((ivy-re-builders-alist '((t . ivy--regex-fuzzy))))
                      (ivy-read "pattern: " '("package-list-packages" "plp")))
                   "plp C-m")
         "plp"))
```

**Comment by abo-abo** [comment]
I've added a fix for these two tests, and also the `counsel-M-x` situation that @tsdh mentioned.  If there's more, please open a new issue, preferably with a failing test.

**Comment by PythonNut** [comment]
@abo-abo awesome. Thanks for your work!
